AutomatedTesting/Gem/PythonTests/Atom/atom_utils/material_canvas_utils.py

```diff
@@ -10,101 +10,101 @@
-    :param source_slot: An GraphModelSlotId object representing the slot on the source node the connection uses.
-    :param target_node: An AZStd::shared_ptr<Node> object representing the target node for the connection.
-    :param target_slot: An GraphModelSlotId object representing the slot on the target node the connection uses.
-    :return: bool representing success (True) or failure (False).
-    """
-    return graph.GraphControllerRequestBus(
-        bus.Event, GraphControllerRequestBusEvents.ARE_SLOTS_CONNECTED, graph_id,
-        source_node, source_slot, target_node, target_slot)
+class Graph(object):
+    """
+    Opens the specified materialgraph file when inside MaterialCanvas which appears as a Graph.
+    Opening the Graph is required since it also sets the document_id for use with other functions.
+    """
+
+    def __init__(self, material_graph_file_path: str):
+        self.material_graph_file_path = material_graph_file_path
+        self.document_id = atom_tools_utils.open_document(self.material_graph_file_path)
+        self.graph_id = atomtools.GraphDocumentRequestBus(
+            bus.Event, GraphDocumentRequestBusEvents.GET_GRAPH_ID, self.document_id)
+        self.name = atomtools.GraphDocumentRequestBus(
+            bus.Event, GraphDocumentRequestBusEvents.GET_GRAPH_NAME, self.document_id)
+        self.graph_object = atomtools.GraphDocumentRequestBus(
+            bus.Event, GraphDocumentRequestBusEvents.GET_GRAPH, self.document_id)
+        self.nodes = []
+
+    def _create_node_by_name(self, node_name: str) -> object:
+        """
+        Creates a new node in memory matching the node_name string on the specified graph object.
+        i.e. "World Position" for node_name would create a World Position node.
+        :param node_name: String representing the type of node to add to the graph.
+        :return: azlmbr.object.PythonProxyObject representing a C++ AZStd::shared_ptr<Node> object.
+        """
+        return atomtools.DynamicNodeManagerRequestBus(
+            bus.Broadcast, DynamicNodeManagerRequestBusEvents.CREATE_NODE_BY_NAME, self.graph_object, node_name)
+
+    def add_node(self, node_name: str, position: math.Vector2) -> None:
+        """
+        Creates a node in memory using _create_node_by_name then adds it to this Graph object.
+        :param node_name: String representing the type of node to add to the graph.
+        :param position: math.Vector2(x,y) value that determines where to place the node on the Graph's grid.
+        :return: None, but appends the new node to self.nodes list.
+        """
+        node_object = self._create_node_by_name(node_name)
+        node_id = graph.GraphControllerRequestBus(
+            bus.Event, GraphControllerRequestBusEvents.ADD_NODE, self.graph_id, node_object, position)
+        self.nodes.append(Node(node_object, node_id, node_name, self.graph_id, position))
```

Comment:
This should verify that the node was created. You should also be able to get the nodes from the graph after all of the functions I exposed. I'm not sure if you can get position but that should be added if you can't.

Reply:
We like to split verification functions from the utils and only verify stuff in the test scripts themselves. If its incorrect it should fail on the test steps.

Reply:
That's fine. I'm just saying that should not be added to the list if it is not a valid node unless you plan to crawl the list and check manually